### ISSUE TITLE
ValueError: Padding is incorrect. With --hls-segment-key-uri option.

### ISSUE THREAD

**Issue by kamikazechaser** [streamlink/streamlink#4528]
### Checklist

- [X] This is a bug report and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed bug reports](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22bug%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest build from the master branch

### Description

Possibly related to #4300, but since that one has been specifically addressed, opening a new one here.

`localhost:8080/key`  returns the unmangled key based on the `_decrypt_key` implementation [here](https://github.com/streamlink/streamlink/issues/2253#issuecomment-455571775). I assume if that was the issue, it would throw an AES/decryption error instead.

> Another problem is that the HLSStreamWriter.fetch method (which calls create_request_params) is run by other threads from the writer's thread-pool, so a sequential order of requests is not guaranteed.

Sounds like this might be the issue. Not sure though.

### Debug log

```text
streamlink -l debug https://vcdn.spayee.in/spees/w/o/607d7f740cf265c8e0215b0d/v/60a0c1310cf20fa1ed1a0bf0/u/626cf77b0cf20486a589e9e8/t/39bca27e4d2ece835d78adaa651a019e/p/assets/videos/607d7f740cf265c8e0215b0d/2021/05/16/60a0c1310cf20fa1ed1a0bf0/index.m3u8 500k --hls-segment-key-uri "http://localhost:8080/key"
[cli][info] Loaded config from deprecated path, see CLI docs for how to migrate: /home/kamikaze/.streamlinkrc
[cli][debug] OS:         Linux-5.10.52-antix.1-amd64-smp-x86_64-with-glibc2.31
[cli][debug] Python:     3.9.2
[cli][debug] Streamlink: 4.0.1+29.g9e4a6577
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.3.2)
[cli][debug] Arguments:
[cli][debug]  url=https://vcdn.spayee.in/spees/w/o/607d7f740cf265c8e0215b0d/v/60a0c1310cf20fa1ed1a0bf0/u/626cf77b0cf20486a589e9e8/t/39bca27e4d2ece835d78adaa651a019e/p/assets/videos/607d7f740cf265c8e0215b0d/2021/05/16/60a0c1310cf20fa1ed1a0bf0/index.m3u8
[cli][debug]  stream=['500k']
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][debug]  --hls-segment-key-uri=http://localhost:8080/key
[cli][info] Found matching plugin hls for URL https://vcdn.spayee.in/spees/w/o/607d7f740cf265c8e0215b0d/v/60a0c1310cf20fa1ed1a0bf0/u/626cf77b0cf20486a589e9e8/t/39bca27e4d2ece835d78adaa651a019e/p/assets/videos/607d7f740cf265c8e0215b0d/2021/05/16/60a0c1310cf20fa1ed1a0bf0/index.m3u8
[plugins.hls][debug] URL=https://vcdn.spayee.in/spees/w/o/607d7f740cf265c8e0215b0d/v/60a0c1310cf20fa1ed1a0bf0/u/626cf77b0cf20486a589e9e8/t/39bca27e4d2ece835d78adaa651a019e/p/assets/videos/607d7f740cf265c8e0215b0d/2021/05/16/60a0c1310cf20fa1ed1a0bf0/index.m3u8; params={}
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 350k (worst), 500k, 1000k (best)
[cli][info] Opening stream: 500k (hls)
[cli][info] Starting player: mpv
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Segments in this playlist are encrypted
[stream.hls][debug] First Sequence: 0; Last Sequence: 1182
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 0; End Sequence: 1182
[stream.hls][debug] Adding segment 0 to queue
[stream.hls][debug] Adding segment 1 to queue
[stream.hls][debug] Adding segment 2 to queue
[stream.hls][debug] Adding segment 3 to queue
[stream.hls][debug] Adding segment 4 to queue
[stream.hls][debug] Adding segment 5 to queue
[stream.hls][debug] Adding segment 6 to queue
[stream.hls][debug] Adding segment 7 to queue
[stream.hls][debug] Adding segment 8 to queue
[stream.hls][debug] Adding segment 9 to queue
[stream.hls][debug] Adding segment 10 to queue
[stream.hls][debug] Adding segment 11 to queue
[stream.hls][debug] Adding segment 12 to queue
[stream.hls][debug] Adding segment 13 to queue
[stream.hls][debug] Adding segment 14 to queue
[stream.hls][debug] Adding segment 15 to queue
[stream.hls][debug] Adding segment 16 to queue
[stream.hls][debug] Adding segment 17 to queue
[stream.hls][debug] Adding segment 18 to queue
[stream.hls][debug] Adding segment 19 to queue
[stream.hls][debug] Adding segment 20 to queue
[stream.hls][debug] Adding segment 21 to queue
Exception in thread Thread-HLSStreamWriter:
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/home/kamikaze/.local/lib/python3.9/site-packages/streamlink/stream/segmented.py", line 200, in run
    self.write(segment, result, *data)
  File "/home/kamikaze/.local/lib/python3.9/site-packages/streamlink/stream/hls.py", line 196, in write
    return self._write(sequence, result, *data)
  File "/home/kamikaze/.local/lib/python3.9/site-packages/streamlink/stream/hls.py", line 231, in _write
    chunk = unpad(decrypted_chunk, AES.block_size, style="pkcs7")
  File "/home/kamikaze/.local/lib/python3.9/site-packages/Crypto/Util/Padding.py", line 92, in unpad
    raise ValueError("Padding is incorrect.")
ValueError: Padding is incorrect.
^CInterrupted! Exiting...
[cli][info] Closing currently open stream...
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
```

**Comment by bastimeyer** [comment]
> ValueError("Padding is incorrect.")

As you can see in the call stack, the HLS stream implementation is calling [`Crypto.Util.Padding.unpad(decrypted_chunk, AES.block_size, style="pkcs7")`](https://github.com/streamlink/streamlink/blob/f08bc5a8adbd9333ed262ed020e9c6c99d6b274f/src/streamlink/stream/hls.py#L231) to unpad the decrypted segment content. Unpadding means that additional data needs to get removed that was added prior to encrypting it, to make it fit exactly into a multiple of the AES block size, which is 16 bytes.

The `ValueError` gets raised by `pycryptodome` [here](https://github.com/Legrandin/pycryptodome/blob/v3.14.1/lib/Crypto/Util/Padding.py#L92) if the PKCS#7 padding format/value of the decrypted segment is invalid. The decrypted segment requires its last byte to indicate the padding size, [according to the PKCS#7 algorithm (RFC 5652)](https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS#5_and_PKCS#7).

The "Padding is incorrect." means that the last byte was either `0x00`, or a value greater than `0x10`, which doesn't make sense in the context of AES block sizes of 16 bytes.

In order to debug this, it would be necessary to see the contents of the decrypted segment. You could modify Streamlink's HLS code and write it to the filesystem before it attempts to unpad it. Since I don't have the decryption key, I can't help with this.

The decrypted segment could also very likely be complete junk data, depending on the key and initialization vector (IV), which would be the reason for the incorrect padding data.

----

#4300 is unrelated to this. #4300 was caused by a faulty HLS byterange implementation after a code refactor in the HLS parser. This has been resolved in the `3.1.0` release.

There was also another issue in the past in regards to incorrect PKCS7 padding, and that was caused by `requests`/`urllib3` not respecting the `Content-Length` header of certain HTTP requests, leading to partial HLS segment downloads which then were obviously missing their PKCS#7 padding bytes at the end. This was fixed by #3768 in the `2.2.0` release.

**Comment by bastimeyer** [comment]
I'm going to close the thread if neither the decryption key nor the contents of the first decrypted segment get posted here, as this issue looks like you're trying to use an invalid decryption key. There won't be any "decryption errors" unless the key itself is malformed (eg. an invalid key length). If the key is wrong, then the result will simply be junk data and the invalid padding error will get raised unless the junk data happens to randomly align with the PKCS#7 padding format.

You can get the encrypted and decrypted+padded content of the first segment by applying this diff:
```diff
diff --git a/src/streamlink/stream/hls.py b/src/streamlink/stream/hls.py
index dedc7bbf..86de78a7 100644
--- a/src/streamlink/stream/hls.py
+++ b/src/streamlink/stream/hls.py
@@ -228,8 +228,13 @@ class HLSStreamWriter(SegmentedStreamWriter):
             else:
                 decrypted_chunk = decryptor.decrypt(data)
 
-            chunk = unpad(decrypted_chunk, AES.block_size, style="pkcs7")
-            self.reader.buffer.write(chunk)
+            log.info(f"Writing data of segment {sequence.num}")
+            with open(f"segment-encrypted-{sequence.num}.ts", "wb") as fh:
+                fh.write(data)
+            with open(f"segment-decrypted-{sequence.num}.ts", "wb") as fh:
+                fh.write(decrypted_chunk)
+
+            self.reader.close()
 
         else:
             try:
```
